src/euroleague_api/team_lineup_euroleague_api.py

```diff
@@ -0,0 +1,98 @@
+            sub_time = pbp_data.loc[idx, 'MARKERTIME']
+            sub_min = pbp_data.loc[idx, 'MINUTE']
+            sub_team = pbp_data.loc[idx, 'TEAM']
+            all_subs = pbp_data.loc[(pbp_data.MINUTE == sub_min) & (pbp_data['MARKERTIME'] == sub_time) & (pbp_data.PLAYTYPE.isin(['IN', 'OUT'])), :].sort_values(by='PLAYTYPE')
+
+            if pd.isnull(pbp_data.loc[max(all_subs.index), 'Team_A1']): # if I haven't made any replacement yet
+                subs_team1 = all_subs.loc[all_subs['TEAM'] == sub_team, :]
+                subs_team2 = all_subs.loc[all_subs['TEAM'] != sub_team, :]
+                in_players_1 = []
+                out_players_1 = []
+                for i in subs_team1.index: # capturing every sub made by the first team at that particular time in the match
+                    if subs_team1.loc[i, 'PLAYTYPE'] == 'IN':
+                        tmp_in_player = subs_team1.loc[i, 'PLAYER']
+                        in_players_1.append(tmp_in_player)
+                    elif subs_team1.loc[i, 'PLAYTYPE'] == 'OUT':
+                        tmp_out_player = subs_team1.loc[i, 'PLAYER']
+                        out_players_1.append(tmp_out_player)
+                
+                in_players_2 = []
+                out_players_2 = []
+                for i in subs_team2.index: # capturing every sub made by the second team at that particular time in the match
+                    if subs_team2.loc[i, 'PLAYTYPE'] == 'IN':
+                        tmp_in_player = subs_team2.loc[i, 'PLAYER']
+                        in_players_2.append(tmp_in_player)
+                    elif subs_team2.loc[i, 'PLAYTYPE'] == 'OUT':
+                        tmp_out_player = subs_team2.loc[i, 'PLAYER']
+                        out_players_2.append(tmp_out_player)
+                
+                subs_dict = {} # empty dictionary to create key-value pairs for the subs
+
+                all_ins = in_players_1 + in_players_2 # every player who subs in
+                all_outs = out_players_1 + out_players_2 # every player who subs out
+
+                for out_p, in_p in zip(all_outs, all_ins):
+                    subs_dict[out_p] = in_p # creating the dictionary appropriately. Player who subs out becomes the key while the player who subs in becomes the value
+                
+                for key in subs_dict.keys(): # make the subs and hold them in the players variable
+                    for j, player in enumerate(players):
+                        if key == player:
+                            players[j] = subs_dict[key]
+                        else:
+                            continue
+                
+                pbp_data.loc[max(all_subs.index), on_court_cols] = players # replace them in the actualy dataframe
+            
+            else:
+                pass
+    
+    pbp_data.loc[:, on_court_cols] = pbp_data.loc[:, on_court_cols].fillna(method='ffill') # filling every play
```

Comment:
Since you are fillining the non-sub rows at the end, why don't you iterate and in the outer for loop only the in/our sub rows i.e. `sub_pbp_data = pbp_data["PLAYTYPE"].isin(["IN", "OUT"]`. It will be much faster.